plugins-builtin/src/main/scala/za/co/absa/enceladus/plugins/builtin/errorsender/mq/KafkaErrorSenderPluginImpl.scala

```diff
@@ -0,0 +1,191 @@
+   * Processes the `dataFrame` - errors are exploded (one error = one line) and filtered to conform to the error source (standardization/conformance)
+   * @param dataFrame standardization/conformance output
+   * @param params plugin processing parameters
+   * @return DF with exploded errors and corresponding to the given error source
+   */
+  def getIndividualErrors(dataFrame: DataFrame, params: ErrorSenderPluginParams): DataFrame = {
+    implicit val singleErrorStardardizedEncoder = Encoders.product[SingleErrorStardardized]
+    implicit val dceErrorEncoder = Encoders.product[DceError]
+
+    val allowedErrorCodes = KafkaErrorSenderPluginImpl.errorCodesForSource(params.sourceId)
+
+    val stdErrors = dataFrame
+      .filter(size(col("errCol")) > 0)
+      // only keep columns that are needed for the actual error publishing
+      .select(
+        col(ColumnNames.enceladusRecordId).cast(DataTypes.StringType).as("recordId"),
+        col(ColumnNames.reportDate),
+        explode(col(ColumnNames.errCol)).as("singleError")
+      )
+      .as[SingleErrorStardardized]
+      .map(_.toDceError(params))
+      .filter(entry => allowedErrorCodes.contains(entry.errorCode)) // Std xor Conf error codes
+      .toDF()
+
+    stdErrors
+  }
+
+  def prepareDataForKafka(stdErrors: DataFrame, params: ErrorSenderPluginParams): DataFrame = {
+    val valueAvroSchemaString = KafkaErrorSenderPlugin.getValueAvroSchemaString
+    val valueSchemaType = KafkaErrorSenderPlugin.getValueStructTypeSchema
+
+    val keyAvroSchemaString = KafkaErrorSenderPlugin.getKeyAvroSchemaString
+
+    val allValueColumns = struct(stdErrors.columns.head, stdErrors.columns.tail: _*)
+    import za.co.absa.abris.avro.functions.to_confluent_avro
+
+    stdErrors.sqlContext.createDataFrame(stdErrors.rdd, valueSchemaType) // forces avsc schema to assure compatible nullability of the DF
+      .select(
+        to_confluent_avro(
+          struct(lit(params.sourceSystem).as("sourceSystem")), keyAvroSchemaString, keySchemaRegistryConfig
+        ).as("key"),
+        to_confluent_avro(allValueColumns, valueAvroSchemaString, valueSchemaRegistryConfig).as("value")
+      )
+  }
+
+  /**
+   * Actual data sending
+   * @param df Dataframe with confluent_avro columns - key & value
+   */
+  private[mq] def sendErrorsToKafka(df: DataFrame): Unit = {
+    require(df.schema.fieldNames.contains("key") && df.schema.fieldNames.contains("value"))
+
+    df.write
+      .format("kafka")
+      .option("kafka.bootstrap.servers", connectionParams.bootstrapServers)
+      .option("topic", connectionParams.topicName)
+      .option("kafka.client.id", connectionParams.clientId)
+      .option("path", "notReallyUsedButAtumExpectsItToBePresent") // TODO Atum issue #32
+      .save()
+  }
+}
+
+object KafkaErrorSenderPluginImpl {
+
+  // columns from the original datafram (post Stdardardization/Conformance) to be addressed
+  object ColumnNames {
+    val enceladusRecordId = "enceladus_record_id"
```

Comment:
I think there should be constants defined somewhere. As well as there is more, isn't there?

Reply:
These are actually the constants you probably mean. And from here, all other usages within the module (`plugins-builtin`) reference this.

The only other place where this is not reused (not counting tests where it is written out deliberately) is `za.co.absa.enceladus.common.Constants` - that is a different module `spark-jobs`.

I am a bit reluctant to reference one another because they serve a slightly different purpose:
the former is a "column that must be present for the plugin to work", the latter is a column that is added when not present to the Std/Conf output. 

Do you believe they should truly reference one another? 🤔

Reply:
I don't anymore